lib/ansible/module_utils/urls.py

```diff
@@ -300,6 +302,51 @@ def http_open(self, req):
         return self.do_open(UnixHTTPConnection(self._unix_socket), req)
 
 
+class ProxyHandler(urllib.request.ProxyHandler):
+    _SPLITPORT_RE = re.compile('(.*):([0-9]*)', re.DOTALL)
+
+    @classmethod
+    def _splitport(cls, host):
+        # Copied from cpython urllib.parse
+        match = cls._SPLITPORT_RE.fullmatch(host)
+        if match:
+            host, port = match.groups()
+            if port:
+                return host, port
+        return host, None
+
+    @staticmethod
+    def _matches(host, port, bypass_network, bypass_port, scheme):
+        if not port:
+            port = '443' if scheme == 'https' else '80'
+        if bypass_port and port != bypass_port:
+            return False
+        return host in bypass_network
+
+    def proxy_open(self, req, *args):
```

Comment:
The biggest concern I have about this implementation, is in overriding this method.  `proxy_open` is not a documented method at https://docs.python.org/3/library/urllib.request.html

And further more, it's highly dependent on `urllib.request.ProxyHandler.__init__`: https://github.com/python/cpython/blob/8f25cc992021d6ffc62bb110545b97a92f7cb295/Lib/urllib/request.py#L770-L774

However, the use of `proxy_open` has been around, unchanged for 25+ years, so it seems relatively safe and stable: https://github.com/python/cpython/commit/6d7e47b8ea1b8cf82927dacc364689b8eeb8708b

I guess at a minimum, this is why we have tests, which this PR includes.